### DISCUSSION TITLE
Question regarding GC Handle

### DISCUSSION THREAD

**Discussion by Flechaa** [vfsfitvnm/frida-il2cpp-bridge#293] [Q&A]
Everytime I create a C# String it gets garbage collected after a while. I am pinning the object so that it does not get GC'ed but I don't know if I really need to free the handle when the script is unloaded like this:

```
const string = Il2Cpp.string("hello");
const gcHandle = string.object.ref(true);
Script.bindWeak(globalThis, () => gcHandle.free());
```

**Comment by vfsfitvnm** [comment]
That sounds technically right. I honestly never faced this situation, and I think the root cause is `Il2Cpp::perform` detaching the caller thread from IL2CPP immediately after executing the given block. So, your thread creates the string, you pass the string somewhere, the thread gets detached, the gc frees the string.

I've recently pushed a [change](https://github.com/vfsfitvnm/frida-il2cpp-bridge/commit/217a9cd3ef62a87456650323d2aa0da5e4c9b43a) so that the Frida thread gets detached from IL2CPP only when the script is about to get unloaded; this will allow the IL2CPP internals to be accessible within the Frida REPL or inside `setTimetout` (& co.) callbacks (right now `Il2Cpp.currentThread` is `null` in there).
This is virtually equivalent to freeing the handle when the script is unloaded.

However, I'm not sure if this change really helps you (imagine this scenario: your thread creates the string, you pass the string somewhere, you reload the script or exit Frida, the gc frees the string).

Perhaps I can add a parameter to `Il2Cpp::perform` so that it never detaches the caller thread. However, I don't know what would happen if the attached thread ends its execution (well, there would be a memory leak but it isn't really a concern).

Alternatively, you could create the string from the main thread instead:
```ts
Il2Cpp.perform(async () => {
    const string = await Il2Cpp.mainThread.schedule(() => Il2Cpp.string("hello"));
});
``` 
However, notice the synchronization context of a thread (this is what `Il2Cpp.Thread::schedule` uses underneath) isn't available during early instrumentation or, in old Unity versions, may not be available at all - so you might have to wrap that code inside a `Promise` + `setTimeout`.

**Comment by vfsfitvnm** [comment]
I honestly don't know if there's a way to keep an object always alive, despite the thread who created it exits or gets detached.
I think your best bet is to create that object from the main thread.

**Comment by vfsfitvnm** [comment]
After a quick read at the documentation, the object won't be collected as long as you create a strong reference to it (pinning is not necessary), regardless of the thread who created it. If you want that string to be always alive (aka leak it), don't free the handle.

https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.gchandle
https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.gchandletype

**Reply by Flechaa** [reply]
Actually I don't mind the string being freed because it will be recreated once the script reloads I just wasn't sure if pinning it and then not freeing it would create a memory leak.

**Reply by vfsfitvnm** [reply]
Got it. You have to leak the object for now (until `v0.8.1` is released).
Keep in mind you still have to leak it (or create it from the main thread) if you want to keep it alive forever.

**Reply by Flechaa** [reply]
So perform finds a random thread to attach to? I have looked over the source code but there's some parts I haven't fully understood.

**Reply by vfsfitvnm** [reply]
`Il2Cpp::perform` attaches the caller thread to IL2CPP. In most cases, the caller thread simply is the Frida's one.
However, in complex scenarios, the caller thread might be different - for instance, the callbacks you provide to `Interceptor.attach` are executed on the thread that is invoking the function you intercepted. If you call `Il2Cpp::perform` from there, the caller thread might be neither the Frida's one nor a IL2CPP one.

**Reply by Flechaa** [reply]
I have done that before when Il2Cpp.delegate() didn't exist but sometimes I would get "abort was called". That was probably a thread issue

**Reply by Flechaa** [reply]
> Got it. You have to leak the object for now (until `v0.8.1` is released). Keep in mind you still have to leak it (or create it from the main thread) if you want to keep it alive forever.

So with the latest commit I do not have to pin the object right? I don't mind if they get garbage collected as I only need it while frida is attached.

**Reply by vfsfitvnm** [reply]
Kind of. If you don't reload the script, then, you don't have to create a strong reference

**Reply by Flechaa** [reply]
I also don't care if it's destroyed once I reload the script so I am fine right?

**Reply by vfsfitvnm** [reply]
Yes
